_posts/2024-11-19-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,623 @@
+
+IGNORE_THESHOLD = 5000.0
+class AboveThresholdAccum(Accumulator):
+    def __init__(self) -> None:
+        self._sum = 0.0
+
+    def update(self, values: pa.Array) -> None:
+        over_threshold = pc.greater(values, pa.scalar(IGNORE_THESHOLD))
+        sum_above = pc.sum(values.filter(over_threshold)).as_py()
+        if sum_above is None:
+            sum_above = 0.0
+        self._sum = self._sum + sum_above
+
+    def merge(self, states: List[pa.Array]) -> None:
+        self._sum = self._sum + pc.sum(states[0]).as_py()
+
+    def state(self) -> List[pa.Scalar]:
+        return [pa.scalar(self._sum)]
+
+    def evaluate(self) -> pa.Scalar:
+        return pa.scalar(self._sum)
+
+sum_above_threshold = udaf(AboveThresholdAccum, [pa.float64()], pa.float64(), [pa.float64()], 'stable')
+
+df_orders.aggregate([col("o_custkey")],[sum_above_threshold(col("o_totalprice")).alias("sales")]).show()
+```
+
+Since we are doing a `sum` we can keep a single value as our internal state. When we call `update()`
+we will process a single array and update the internal state, which we share with the `state()`
+function. For larger batches we may `merge()` these states. It is important to note that the
+`states` in the `merge()` function are an array of the values returned from `state()`. It is
+entirely possible that the `merge` function is significantly different than the `update`, though in
+our example they are very similar.
+
+## User Defined Window Functions
+
+Writing a user defined window function is slighlty more complex than an aggregate function due
+to the variety of ways that window functions are called. I recommend reviewing the
+[online documentation](https://datafusion.apache.org/python/user-guide/common-operations/udf-and-udfa.html)
+for a description of which functions need to be implemented. The details of how to implement
+these generally follow the same patterns as described above for aggregate functions.
+
+## Performance Comparison
+
+For the scalar functions above, we performed a timing evaluation, repeating the operation 100
+times. For this simple example these are our results.
+
+```
++-----------------------------+--------------+---------+
+| approach                    | Average Time | Std Dev |
++-----------------------------+--------------+---------+
+| python udf                  | 4.969        | 0.062   |
+| simple filter               | 1.075        | 0.022   |
+| explicit filter             | 0.685        | 0.063   |
+| pyarrow compute             | 0.529        | 0.017   |
+| arrow rust compute          | 0.511        | 0.034   |
+| arrow rust compute as class | 0.502        | 0.011   |
+| rust custom iterator        | 0.478        | 0.009   |
++-----------------------------+--------------+---------+
+```
+
+As expected, the conversion to Python objects is by far the worst performance. As soon as we drop
+into using any functions that keep the data entirely on the Rust side we see a near 10x speed
```

Comment:
I think technically using pyarrow doesn't use rust -- it is implemented in C/C++. Possibly this would be slightly more accurate:

```suggestion
into using any functions that keep the data entirely on the Native (Rust or C/C++) side we see a near 10x speed
```